### PR TITLE
add sycl KI for InferenceBuilder by using syclomatic

### PR DESCRIPTION
**Pull request by baodii**
* as hipify, we only support JIT mode
* we only support inference kernels now
* delete other OpBuilders in op_builder/xpu/__init__.py and accelerator/xpu_accelerator.py
* add rule.YAML, pre_process.sh and post_process.sh to help us justify source code and generated code
* we don't change original cuda code. we will copy original cuda code to build folder
* sycl code will be generated in third-party folder
* sycl code will be generated but not compiled when install deepspeed as hipify

### CODE REVIEW

**Comment by baodii** [comment]
@delock @CaoZhongZ @rogerxfeng8 please review

File: op_builder/xpu/transformer_inference.py [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+
+    def is_compatible(self, verbose=True):
+        try:
+            import torch
+        except ImportError:
+            self.warning("Please install torch if trying to pre-compile inference kernels")
+            return False
+
+        cuda_okay = True
+        if not self.is_rocm_pytorch() and not self.is_sycl_enabled() and torch.cuda.is_available():
```
Comment (delock) [inline op_builder/xpu/transformer_inference.py]: why there is changes for cuda and rocm?

File: op_builder/xpu/builder.py [hunk 2]
```diff
@@ -6,9 +6,11 @@
 import os
 import time
 import importlib
+import shutil
+from pathlib import Path
+import subprocess
 from deepspeed.ops.op_builder.builder import OpBuilder, TORCH_MAJOR, TORCH_MINOR
 
-
 class SYCLOpBuilder(OpBuilder):
```
Comment (delock) [inline op_builder/xpu/builder.py]: shouldn't there be two kinds of builder, SYCLOpBuilder and SYCLAutoOpBuilder, and SYCLAutoOpBuilder works for ops with syclomatic?

File: op_builder/xpu/transformer_inference.py (hunk 1 above)
Comment (baodii) [reply op_builder/xpu/transformer_inference.py]: already deleted.

File: op_builder/xpu/builder.py (hunk 2 above)
Comment (baodii) [reply op_builder/xpu/builder.py]: > shouldn't there be two kinds of builder, SYCLOpBuilder and SYCLAutoOpBuilder, and SYCLAutoOpBuilder works for ops with syclomatic?

add SYCLAutoOpBuilder below.

File: accelerator/xpu_accelerator.py [hunk 3]
```diff
@@ -242,18 +242,12 @@ def get_op_builder(self, class_name):
...
-            return AsyncIOBuilder
-        elif class_name == "CPUAdagradBuilder":
-            return CPUAdagradBuilder
-        elif class_name == "CPUAdamBuilder":
-            return CPUAdamBuilder
-        elif class_name == "FusedAdamBuilder":
-            return FusedAdamBuilder
+            from deepspeed.ops.op_builder.xpu import CPUAdagradBuilder, CPUAdamBuilder, FusedAdamBuilder, AsyncIOBuilder, InferenceBuilder
+
+        if class_name == "InferenceBuilder":
```
Comment (delock) [inline accelerator/xpu_accelerator.py]: This change seems to turn off CPUAdagradBuilder, CPUAdamBuilder, FusedAdamBuilder, AsyncIOBuilder.  Is this intended?

File: .gitignore [hunk 4]
```diff
@@ -8,6 +8,7 @@ __pycache__
 
 # Build + installation data
 build/
+third-party/
```
Comment (delock) [inline .gitignore]: Is third-party/ used to store generated sycl kernels?  What will the directory structure be under third-party?

File: op_builder/xpu/post_process.sh [hunk 5]
```diff
@@ -0,0 +1,219 @@
...
+find ./deepspeed/third-party -type f -exec sed -i "s/group_local_memory</group_local_memory_for_overwrite</g" {} +
+
+# fix attn_softmax_v2 lacking of iterations
+find ./deepspeed/third-party/ -type f -exec sed -i "s/attn_softmax_v2<T>/attn_softmax_v2<T, iterations>/g" {} +
+
+# fix device at::kCUDA to at::kXPU
+find ./deepspeed/third-party/ -type f -exec sed -i "s/at::kCUDA/at::kXPU/g" {} +
+
+# fix pt_binding.cpp torch::from_blob 4 inputs pattern
+patch ./deepspeed/third-party/csrc/transformer/inference/csrc/pt_binding.cpp << 'DIFF___'
```
Comment (delock) [inline op_builder/xpu/post_process.sh]: What is the error intend to fix?  Why can't it be fixed in the source code directory? Is it temporary or will persist?  What do we need to do when pt_binding.cpp get changed?

File: op_builder/xpu/builder.py [hunk 6]
```diff
@@ -117,3 +119,247 @@ def jit_load(self, verbose=True):
...
+                                   sources=self.strip_empty_entries(sycl_sources),
+                                   include_dirs=self.strip_empty_entries(sycl_include_paths),
+                                   extra_compile_args={
+                                       'cxx': self.strip_empty_entries(self.cxx_args()),
+                                   },
+                                   extra_link_args=self.strip_empty_entries(self.fixed_aotflags()))
+        return dpcpp_ext
+
+    def sycl_extension(self):
+        if self.is_sycl_enabled():
```
Comment (delock) [inline op_builder/xpu/builder.py:187]: This function is very long.  Can we extract two smaller functions?  one for include, one for source.

File: op_builder/xpu/builder.py [hunk 7]
```diff
@@ -117,3 +119,247 @@ def jit_load(self, verbose=True):
...
+                    sycl_kernel_name = source.replace('.cu', '.dp.cpp')
+                    if os.path.exists(os.path.join(sycl_link_path, sycl_kernel_name)):
+                        print(f'skip migrate {os.path.join(sycl_link_path, sycl_kernel_name)}, we already have one.')
+                        continue
+
+                    need_post_process = True
+                    trans_cmd = c2s_cmd + cuda_inc_flag + extra_args + in_root + out_root + cuda_source
+                    print("**** processing ", f'{trans_cmd}')
+                    p = subprocess.Popen(f'{trans_cmd}', stdout=subprocess.PIPE, shell=True)
+                    # processes_running.append(p)
```
Comment (delock) [inline op_builder/xpu/builder.py]: Please remove code that is not needed.

File: op_builder/xpu/pre_process.sh [hunk 8]
```diff
@@ -0,0 +1,50 @@
+
+# disable PTX_AVAILABLE
+find ./build/csrc -name "*.h" -exec sed -Ei "s:#define.*PTX_AVAILABLE:// \0:g" {} +
+
+# migrate torch::from_blob to at::from_blob
+find ./build/csrc -type f -exec sed -i "s/torch::from_blob/at::from_blob/g" {} +
+
+# fix inference_context.h to make it could be migrate
+patch ./build/csrc/transformer/inference/includes/inference_context.h << 'DIFF___'
```
Comment (delock) [inline op_builder/xpu/pre_process.sh:6]: Why can't change source code directly?

**Comment by delock** [comment]
@baodii  comments added.  Please also fix format error by turning on pre-commit in your environment.

**Comment by baodii** [comment]
> @baodii comments added. Please also fix format error by turning on pre-commit in your environment.

some license check in csrc/xpu folder not pass.

File: op_builder/xpu/pre_process.sh (hunk 8 above)
Comment (baodii) [reply op_builder/xpu/pre_process.sh:6]: > Why can't change source code directly?

add changes in source codes.

**Comment by baodii** [comment]
@delock I have fixed these issues. Please review.

**Pull request merged**